File: mrtdreader/src/main/java/com/appliedrec/mrtdreader/MRTDReader.java
package com.appliedrec.mrtdreader;

import static org.jmrtd.PassportService.DEFAULT_MAX_BLOCKSIZE;
import static org.jmrtd.PassportService.NORMAL_MAX_TRANCEIVE_LENGTH;

import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.nfc.tech.IsoDep;

import net.sf.scuba.smartcards.CardFileInputStream;
import net.sf.scuba.smartcards.CardService;

import org.jmrtd.BACKey;
import org.jmrtd.PassportService;
import org.jmrtd.lds.CardSecurityFile;
import org.jmrtd.lds.LDSFileUtil;
import org.jmrtd.lds.PACEInfo;
import org.jmrtd.lds.SODFile;
import org.jmrtd.lds.SecurityInfo;
import org.jmrtd.lds.icao.COMFile;
import org.jmrtd.lds.icao.DG1File;
import org.jmrtd.lds.icao.DG2File;
import org.jmrtd.lds.icao.MRZInfo;
import org.jmrtd.lds.iso19794.FaceImageInfo;
import org.jmrtd.lds.iso19794.FaceInfo;
import org.spongycastle.jce.provider.BouncyCastleProvider;

import java.io.ByteArrayOutputStream;
import java.io.InputStream;
import java.security.Security;
import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

import io.reactivex.rxjava3.core.Observable;
import jj2000.JJ2000Frontend;

public class MRTDReader {

    @FunctionalInterface
    interface Transformer<T, R> {
        R apply(T t) throws Exception;
    }

    private static <T> T readFile(PassportService passportService, short fid) throws Exception {
        try (CardFileInputStream inputStream = passportService.getInputStream(fid)) {
            return (T) LDSFileUtil.getLDSFile(fid, inputStream);
        }
    }

    public static Observable<MRTDReaderProgress> createProgressObservable(IsoDep isoDep, BACSpec bacSpec) {
        BACKey bacKey = new BACKey(bacSpec.getDocumentNumber(), bacSpec.getDateOfBirth(), bacSpec.getDateOfExpiry());
        return Observable.create(emitter -> {
            short fid = PassportService.EF_COM;
            CardService cardService = null;
            PassportService service = null;
            try {

                Security.insertProviderAt(new BouncyCastleProvider(), 1);
                Security.addProvider(new SecurityProvider("MRTDSecurityProvider", 1, "null"));

                cardService = CardService.getInstance(isoDep);
                cardService.open();

                service = new PassportService(cardService, NORMAL_MAX_TRANCEIVE_LENGTH, DEFAULT_MAX_BLOCKSIZE, true, false);
                service.open();

                boolean paceSucceeded = false;
                try (InputStream cardServiceInputStream = service.getInputStream(PassportService.EF_CARD_SECURITY, DEFAULT_MAX_BLOCKSIZE)) {
                    CardSecurityFile cardAccessFile = new CardSecurityFile(cardServiceInputStream);
                    Collection<SecurityInfo> securityInfos = cardAccessFile.getSecurityInfos();
                    for (SecurityInfo securityInfo : securityInfos) {
                        if (securityInfo instanceof PACEInfo) {
                            PACEInfo paceInfo = (PACEInfo)securityInfo;
                            service.doPACE(bacKey, paceInfo.getObjectIdentifier(), PACEInfo.toParameterSpec(paceInfo.getParameterId()), null);
                            paceSucceeded = true;
                        }
                    }
                } catch (Exception e) {
                    e.printStackTrace();
                }

                service.sendSelectApplet(paceSucceeded);

                if (!paceSucceeded) {
                    try (InputStream inputStream = service.getInputStream(PassportService.EF_COM, DEFAULT_MAX_BLOCKSIZE)) {
                        inputStream.read();
                    } catch (Exception e) {
                        service.doBAC(bacKey);
                    }
                }

                MRTDReaderProgress progress = new MRTDReaderProgress();

                COMFile comFile = readFile(service, fid);
                progress.setFileId(fid);
                progress.setProgress(0.25);
                if (!emitter.isDisposed()) {
                    emitter.onNext(progress);
                }

                fid = PassportService.EF_SOD;
                SODFile sodFile = readFile(service, fid);
                progress.setFileId(fid);
                progress.setProgress(0.5);
                if (!emitter.isDisposed()) {
                    emitter.onNext(progress);
                }

                fid = PassportService.EF_DG1;
                DG1File dg1File = readFile(service, fid);
                MRZInfo mrzInfo = dg1File.getMRZInfo();
                progress.getResult().setDateOfBirth(mrzInfo.getDateOfBirth());
                progress.getResult().setDateOfExpiry(mrzInfo.getDateOfExpiry());
                progress.getResult().setDocumentNumber(mrzInfo.getDocumentNumber());
                progress.getResult().setDocumentCode(mrzInfo.getDocumentCode());
                progress.getResult().setGender(mrzInfo.getGender().name());
                progress.getResult().setIssuingState(mrzInfo.getIssuingState());
                progress.getResult().setNationality(mrzInfo.getNationality());
                progress.getResult().setPersonalNumber(mrzInfo.getPersonalNumber());
                progress.getResult().setPrimaryIdentifier(mrzInfo.getPrimaryIdentifier());
                progress.getResult().setSecondaryIdentifiers(mrzInfo.getSecondaryIdentifierComponents());
                progress.setFileId(fid);
                progress.setProgress(0.75);
                if (!emitter.isDisposed()) {
                    emitter.onNext(progress);
                }

                fid = PassportService.EF_DG2;
                DG2File dg2File = readFile(service, fid);
                progress.setFileId(fid);
                if (!emitter.isDisposed()) {
                    emitter.onNext(progress);
                }

                List<FaceImageInfo> allFaceImageInfos = new ArrayList<>();
                List<FaceInfo> faceInfos = dg2File.getFaceInfos();
                for (FaceInfo faceInfo : faceInfos) {
                    allFaceImageInfos.addAll(faceInfo.getFaceImageInfos());
                }

                if (!allFaceImageInfos.isEmpty()) {
                    FaceImageInfo faceImageInfo = allFaceImageInfos.iterator().next();

                    int imageLength = faceImageInfo.getImageLength();
                    try (InputStream inputStream = faceImageInfo.getImageInputStream()) {
                        try (ByteArrayOutputStream outputStream = new ByteArrayOutputStream()) {
                            int read;
                            byte[] buffer = new byte[128];
                            int totalRead = 0;
                            while ((read = inputStream.read(buffer, 0, buffer.length)) > 0) {
                                outputStream.write(buffer, 0, read);
                                totalRead += read;
                                if (!emitter.isDisposed()) {
                                    progress.setProgress(0.75 + (double)totalRead / (double)imageLength * 0.25);
                                    emitter.onNext(progress);
                                }
                            }
                            outputStream.flush();
                            byte[] imageData = outputStream.toByteArray();
                            Bitmap faceBitmap = JJ2000Frontend.decode(imageData);

                            if (faceBitmap != null) {
                                progress.getResult().setFaceImage(faceBitmap);
                            } else {
                                faceBitmap = BitmapFactory.decodeByteArray(imageData, 0, imageData.length);
                                if (faceBitmap != null) {
                                    progress.getResult().setFaceImage(faceBitmap);
                                }
                            }
                        }
                    }
//                    DataInputStream dataInputStream = new DataInputStream(faceImageInfo.getImageInputStream());
//                    byte[] imgData = new byte[imageLength];
//                    dataInputStream.readFully(imgData, 0, imageLength);
//
//                    Bitmap faceBitmap = JJ2000Frontend.decode(imgData);
//
//                    if (faceBitmap != null) {
//                        progress.getResult().setFaceImage(faceBitmap);
//                    } else {
//                        faceBitmap = BitmapFactory.decodeByteArray(imgData, 0, imgData.length);
//                        if (faceBitmap != null) {
//                            progress.getResult().setFaceImage(faceBitmap);
//                        }
//                    }
                }
                if (!emitter.isDisposed()) {
                    progress.setProgress(1.0);
                    emitter.onNext(progress);
                    emitter.onComplete();
                }
            } catch (Exception e) {
                if (!emitter.isDisposed()) {
                    emitter.onError(e);
                }
            } finally {
                if (service != null) {
                    service.close();
                }
                if (cardService != null) {
                    cardService.close();
                }
            }
        });
    }

//    public static BiFunction<MRTDScanResult, ? super MRTDReaderProgress, MRTDScanResult> createResultBiFunction() {
//        return (result, progress) -> {
//            Object file = progress.getFile();
//            if (file instanceof DG1File) {
//                MRZInfo mrzInfo = ((DG1File)file).getMRZInfo();
//                result.setDateOfBirth(mrzInfo.getDateOfBirth());
//                result.setDateOfExpiry(mrzInfo.getDateOfExpiry());
//                result.setDocumentNumber(mrzInfo.getDocumentNumber());
//                result.setDocumentCode(mrzInfo.getDocumentCode());
//                result.setGender(mrzInfo.getGender().name());
//                result.setIssuingState(mrzInfo.getIssuingState());
//                result.setNationality(mrzInfo.getNationality());
//                result.setPersonalNumber(mrzInfo.getPersonalNumber());
//                result.setPrimaryIdentifier(mrzInfo.getPrimaryIdentifier());
//                result.setSecondaryIdentifiers(mrzInfo.getSecondaryIdentifierComponents());
//            } else if (file instanceof DG2File) {
//                List<FaceImageInfo> allFaceImageInfos = new ArrayList<>();
//                List<FaceInfo> faceInfos = ((DG2File)file).getFaceInfos();
//                for (FaceInfo faceInfo : faceInfos) {
//                    allFaceImageInfos.addAll(faceInfo.getFaceImageInfos());
//                }
//
//                if (!allFaceImageInfos.isEmpty()) {
//                    FaceImageInfo faceImageInfo = allFaceImageInfos.iterator().next();
//
//                    int imageLength = faceImageInfo.getImageLength();
//                    DataInputStream dataInputStream = new DataInputStream(faceImageInfo.getImageInputStream());
//                    byte[] imgData = new byte[imageLength];
//                    dataInputStream.readFully(imgData, 0, imageLength);
//
//                    Bitmap faceBitmap = JJ2000Frontend.decode(imgData);
//
//                    if (faceBitmap != null) {
//                        result.setFaceImage(faceBitmap);
//                    } else {
//                        faceBitmap = BitmapFactory.decodeByteArray(imgData, 0, imgData.length);
//                        if (faceBitmap != null) {
//                            result.setFaceImage(faceBitmap);
//                        }
//                    }
//                }
//            }
//            return result;
//        };
//    }
}

File: mrtdreader/src/main/java/com/appliedrec/mrtdreader/MRTDReader.kt
package com.appliedrec.mrtdreader

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.nfc.tech.IsoDep
import jj2000.JJ2000Frontend
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.withContext
import net.sf.scuba.smartcards.CardService
import org.bouncycastle.jce.provider.BouncyCastleProvider
import org.jmrtd.BACKey
import org.jmrtd.PACEKeySpec
import org.jmrtd.PassportService
import org.jmrtd.lds.CardSecurityFile
import org.jmrtd.lds.LDSFileUtil
import org.jmrtd.lds.PACEInfo
import org.jmrtd.lds.SODFile
import org.jmrtd.lds.icao.DG1File
import org.jmrtd.lds.icao.DG2File
import org.jmrtd.lds.iso19794.FaceImageInfo
import java.io.ByteArrayOutputStream
import java.security.Security
import java.security.Signature
import java.security.cert.X509Certificate
import java.util.concurrent.CompletableFuture


internal object MRTDReader {

    @JvmStatic
    @Throws(Exception::class)
    suspend fun readPassport(isoDep: IsoDep, bacSpec: BACSpec, onProgress: ((Progress) -> Unit)?, masterListFuture: CompletableFuture<Collection<X509Certificate>>?): MRTDScanResult = coroutineScope {
        val bacKey = BACKey(bacSpec.documentNumber, bacSpec.dateOfBirth, bacSpec.dateOfExpiry)
        var passportService: PassportService? = null
        try {
            val tasks = arrayOf(
                Progress(0.0, "Authenticating"),
                Progress(0.0, "Reading personal information"),
                Progress(0.0, "Reading face image"),
                Progress(0.0, "Verifying document signature")
            )
            Security.addProvider(BouncyCastleProvider())
            val cardService = CardService.getInstance(isoDep)
            passportService = PassportService(
                cardService,
                PassportService.NORMAL_MAX_TRANCEIVE_LENGTH,
                PassportService.DEFAULT_MAX_BLOCKSIZE,
                true,
                false
            )
            reportProgress(onProgress, tasks, 0)
            passportService.open()
            val paceSucceeded = doPACE(passportService, bacKey)
            passportService.sendSelectApplet(paceSucceeded)
            if (!paceSucceeded) {
                passportService.doBAC(bacKey)
            }
            val sodFile: SODFile = readFile(passportService, PassportService.EF_SOD)
            tasks[0].completed = 1.0
            reportProgress(onProgress, tasks, 1)
            val dG1File: DG1File = readFile(passportService, PassportService.EF_DG1)
            val result = MRTDScanResult.Success(
                documentCode = dG1File.mrzInfo.documentCode,
                issuingState = dG1File.mrzInfo.issuingState,
                primaryIdentifier = dG1File.mrzInfo.primaryIdentifier,
                secondaryIdentifiers = dG1File.mrzInfo.secondaryIdentifierComponents,
                nationality = dG1File.mrzInfo.nationality,
                documentNumber = dG1File.mrzInfo.documentNumber,
                personalNumber = dG1File.mrzInfo.personalNumber,
                dateOfBirth = dG1File.mrzInfo.dateOfBirth,
                dateOfExpiry = dG1File.mrzInfo.dateOfExpiry,
                gender = dG1File.mrzInfo.genderCode.name,
                faceImage = null
            )
            result.dateOfBirth = dG1File.mrzInfo.dateOfBirth
            result.dateOfExpiry = dG1File.mrzInfo.dateOfExpiry
            result.documentNumber = dG1File.mrzInfo.documentNumber
            result.documentCode = dG1File.mrzInfo.documentCode
            result.gender = dG1File.mrzInfo.gender.name
            result.issuingState = dG1File.mrzInfo.issuingState
            result.nationality = dG1File.mrzInfo.nationality
            result.personalNumber = dG1File.mrzInfo.personalNumber
            result.primaryIdentifier = dG1File.mrzInfo.primaryIdentifier
            result.secondaryIdentifiers = dG1File.mrzInfo.secondaryIdentifierComponents
            tasks[1].completed = 1.0
            reportProgress(onProgress, tasks, 2)
            val dG2File: DG2File = readFile(passportService, PassportService.EF_DG2)
            val faceImageInfos = mutableListOf<FaceImageInfo>()
            dG2File.faceInfos.forEach { faceInfo ->
                faceImageInfos.addAll(faceInfo.faceImageInfos)
            }
            if (faceImageInfos.isNotEmpty()) {
                var faceImage: Bitmap?
                for (faceImageInfo in faceImageInfos) {
                    val imageSize = faceImageInfo.imageLength.toDouble()
                    faceImageInfo.imageInputStream.use { inputStream ->
                        ByteArrayOutputStream().use { outputStream ->
                            var read: Int
                            val buffer = ByteArray(1024)
                            var totalRead = 0.0
                            while (inputStream.read(buffer).also { read = it } > 0) {
                                outputStream.write(buffer, 0, read)
                                totalRead += read.toDouble()
                                tasks[2].completed = totalRead / imageSize
                                reportProgress(onProgress, tasks, 2)
                            }
                            outputStream.flush()
                            val imageData = outputStream.toByteArray()
                            faceImage = JJ2000Frontend.decode(imageData)
                            if (faceImage == null) {
                                faceImage =
                                    BitmapFactory.decodeByteArray(imageData, 0, imageData.size)
                            }
                        }
                    }
                    if (faceImage != null) {
                        result.faceImage = faceImage
                        break
                    }
                }
            }
            tasks[2].completed = 1.0
            reportProgress(onProgress, tasks, 3)
            result.signatureVerified = checkDocumentSignature(sodFile)
            tasks[3].completed = 0.3
            reportProgress(onProgress, tasks, 3)
            if (masterListFuture != null) {
                try {
                    val masterList = masterListFuture.get()
                    result.issuerVerified =
                        verifySigningCertificate(sodFile.docSigningCertificate, masterList)
                } catch (e: Exception) {
                    result.issuerVerified = false
                }
            }
            tasks[3].completed = 1.0
            reportProgress(onProgress, tasks, 3)
            result
        } catch (e: Exception) {
            MRTDScanResult.Failure(e)
        } finally {
            passportService?.close()
        }
    }

    private suspend fun reportProgress(onProgress: ((Progress) -> Unit)?, tasks: Array<Progress>, currentTaskIndex: Int) {
        onProgress?.let { consumer ->
            val totalProgress = tasks.sumOf { it.completed } / tasks.size.toDouble()
            val progress = Progress(totalProgress, tasks[currentTaskIndex].message)
            withContext(Dispatchers.Main) {
                consumer(progress)
            }
        }
    }

    private fun checkDocumentSignature(sodFile: SODFile): Boolean {
        try {
            var digestEncryptionAlgorithm = sodFile.digestEncryptionAlgorithm
            if (digestEncryptionAlgorithm == null) {
                return false
            }
            if (digestEncryptionAlgorithm == "SSAwithRSA/PSS") {
                digestEncryptionAlgorithm = sodFile.signerInfoDigestAlgorithm.replace("-", "") + "withRSA/PSS"
            }
            if (digestEncryptionAlgorithm == "RSA") {
                digestEncryptionAlgorithm = sodFile.signerInfoDigestAlgorithm.replace("-", "") + "withRSA"
            }
            var signature: Signature
            try {
                signature = Signature.getInstance(digestEncryptionAlgorithm)
            } catch (e: Exception) {
                signature = Signature.getInstance(digestEncryptionAlgorithm, "BC")
            }
            signature.initVerify(sodFile.docSigningCertificate)
            signature.update(sodFile.eContent)
            return signature.verify(sodFile.encryptedDigest)
        } catch (e: Exception) {
            return false
        }
    }

    private fun <T> readFile(passportService: PassportService, id: Short): T {
        passportService.getInputStream(id, PassportService.DEFAULT_MAX_BLOCKSIZE).use { inputStream ->
            return LDSFileUtil.getLDSFile(id, inputStream) as T
        }
    }

    private fun verifySigningCertificate(certificate: X509Certificate, masterList: Collection<X509Certificate>): Boolean {
        val issuers = masterList.filter { it.subjectX500Principal == certificate.issuerX500Principal }
        for (issuer in issuers) {
            try {
                certificate.verify(issuer.publicKey)
                return true
            } catch (ignore: Exception) {
            }
        }
        return false
    }

    private fun doChipAuthentication(passportService: PassportService): Boolean {
        return false
//        try {
//            val dg14File: DG14File = readFile(passportService, PassportService.EF_DG14)
//            dg14File.securityInfos.firstOrNull { it is ChipAuthenticationPublicKeyInfo }?.let { securityInfo ->
//                val publicKey = (securityInfo as ChipAuthenticationPublicKeyInfo).subjectPublicKey
//                val keyId = securityInfo.keyId
//                val agreementAlgo = Util.inferKeyAgreementAlgorithm(publicKey)
//                val keyPairGenerator = KeyPairGenerator.getInstance(agreementAlgo)
//                var params: AlgorithmParameterSpec? = null
//                if ("DH" == agreementAlgo) {
//                    val dhPublicKey: DHPublicKey = publicKey as DHPublicKey
//                    params = dhPublicKey.getParams()
//                } else if ("ECDH" == agreementAlgo) {
//                    val ecPublicKey: ECPublicKey = publicKey as ECPublicKey
//                    params = ecPublicKey.getParams()
//                } else {
//                    throw IllegalStateException("Unsupported algorithm \"$agreementAlgo\"")
//                }
//                keyPairGenerator.initialize(params)
//                val keyPair = keyPairGenerator.generateKeyPair()
//                val keyAgreement = KeyAgreement.getInstance(agreementAlgo)
//                keyAgreement.init(keyPair.private)
//                keyAgreement.doPhase(publicKey, true)
//                val secret = keyAgreement.generateSecret()
//                var keyData: ByteArray = ByteArray(0)
//                var idData: ByteArray = ByteArray(0)
//                var keyHash: ByteArray = ByteArray(0)
//                if ("DH" == agreementAlgo) {
//                    val dhPublicKey = keyPair.public as DHPublicKey
//                    keyData = dhPublicKey.y.toByteArray()
//                    val md = MessageDigest.getInstance("SHA1")
//                    keyHash = md.digest(keyData)
//                } else {
//                    val ecPublicKey = keyPair.public as org.spongycastle.jce.interfaces.ECPublicKey
//                    keyData = ecPublicKey.q.getEncoded(false)
//                    val t = Util.i2os(ecPublicKey.q.x.toBigInteger())
//                    keyHash = Util.alignKeyDataToSize(t, ecPublicKey.parameters.curve.fieldSize / 8)
//                }
//                keyData = Util.wrapDO(0x91.toByte(), keyData)
//                if (keyId.compareTo(BigInteger.ZERO) >= 0) {
//                    val keyIdBytes = keyId.toByteArray()
//                    idData = Util.wrapDO(0x84.toByte(), keyIdBytes)
//                }
//                passportService.sendMSEKAT(passportService.getWrapper(), keyData, idData)
//
//                val ksEnc = Util.deriveKey(secret, Util.ENC_MODE)
//                val ksMac = Util.deriveKey(secret, Util.MAC_MODE)
//
//                passportService.setWrapper(DESedeSecureMessagingWrapper(ksEnc, ksMac, 0L))
//                val fld: Field = PassportService::class.java.getDeclaredField("state")
//                fld.setAccessible(true)
//                fld.set(passportService, 4) //PassportService.CA_AUTHENTICATED_STATE)
//
//                return ChipAuthenticationResult(keyId, publicKey, keyHash, keyPair)
//            }
//        } catch (ignore: Exception) {}
    }

    private fun doPACE(passportService: PassportService, bacKey: BACKey): Boolean {
        try {
            passportService.getInputStream(
                PassportService.EF_CARD_SECURITY,
                PassportService.DEFAULT_MAX_BLOCKSIZE
            ).use { inputStream ->
                val cardAccessFile = CardSecurityFile(inputStream)
                val securityInfo =
                    cardAccessFile.securityInfos.firstOrNull { it is PACEInfo } as? PACEInfo
                if (securityInfo != null) {
                    passportService.doPACE(
                        PACEKeySpec.createMRZKey(bacKey),
                        securityInfo.objectIdentifier,
                        PACEInfo.toParameterSpec(securityInfo.parameterId),
                        null
                    )
                    return true
                }
            }
        } catch (ignore: Exception) {}
        return false
    }
}